File: api/app/records.py
"""Personal records: whether a set just logged beat everything before it.

Three kinds, checked against every earlier working set of the exercise:

- **weight**: heavier than ever.
- **e1rm**: a better estimated one-rep max, by the same Epley formula as
  v_sets_enriched, so the number matches the Trends chart.
- **reps**: more reps than any set at this weight or heavier. Only where such a
  set exists: at a weight never lifted before, the weight record says it.

A first set is no record; there is nothing to beat. Warmups are neither
checked nor counted. Specified case by case in fixtures/records.json, which
the Kotlin port is held to as well.
"""

from __future__ import annotations

from dataclasses import dataclass

from sqlalchemy import or_, select
from sqlalchemy.orm import Session

from app.models import Exercise, SetEntry, Workout


@dataclass(frozen=True)
class Record:
    kind: str  # "weight" | "e1rm" | "reps"
    value: float
    previous: float

    def as_dict(self) -> dict:
        return {"kind": self.kind, "value": self.value, "previous": self.previous}


def e1rm(weight: float | None, reps: int | None) -> float | None:
    if weight is None or weight <= 0 or reps is None or not 1 <= reps <= 12:
        return None
    return round(weight * (1 + reps / 30.0), 2)


def beaten(before: list[tuple[float | None, int | None]], new: tuple[float | None, int | None]):
    """The records `new` sets over `before`, in the order weight, e1rm, reps."""
    if not before or new[1] is None:
        return []
    weight, reps = new[0] or 0.0, new[1]
    found: list[Record] = []

    heaviest = max((w or 0.0) for w, _ in before)
    if weight > 0 and weight > heaviest:
        found.append(Record("weight", weight, heaviest))

    mine = e1rm(new[0], reps)
    estimates = [e for e in (e1rm(w, r) for w, r in before) if e is not None]
    if mine is not None and estimates and mine > max(estimates):
        found.append(Record("e1rm", mine, max(estimates)))

    at_or_above = [r or 0 for w, r in before if (w or 0.0) >= weight]
    if at_or_above and reps > max(at_or_above):
        found.append(Record("reps", reps, max(at_or_above)))
    return found


def for_set(session: Session, set_id: int) -> tuple[str, list[Record]] | None:
    """The exercise's name and the records this set broke, or None if no such set.

    "Earlier" is an earlier day, or the same day and logged first, so a record
    set earlier in the session is the one the next set has to beat.
    """
    row = session.execute(
        select(SetEntry, Workout.performed_on, Exercise.name)
        .join(Workout, Workout.id == SetEntry.workout_id)
        .join(Exercise, Exercise.id == SetEntry.exercise_id)
        .where(SetEntry.id == set_id)
    ).first()
    if row is None:
        return None
    entry, day, name = row
    if entry.is_warmup:
        return name, []

    before = session.execute(
        select(SetEntry.weight_kg, SetEntry.reps)
        .join(Workout, Workout.id == SetEntry.workout_id)
        .where(
            SetEntry.exercise_id == entry.exercise_id,
            SetEntry.is_warmup.is_(False),
            SetEntry.id != set_id,
            or_(
                Workout.performed_on < day,
                (Workout.performed_on == day) & (SetEntry.id < set_id),
            ),
        )
    ).all()

    def weight(value) -> float | None:
        return float(value) if value is not None else None

    return name, beaten(
        [(weight(w), r) for w, r in before], (weight(entry.weight_kg), entry.reps)
    )

File: android/core/src/main/kotlin/dev/aura/core/Records.kt
package dev.aura.core

import kotlin.math.round

/** A personal record one set broke: what kind, the new best, and the old one. */
data class Record(val kind: Kind, val value: Double, val previous: Double) {
    enum class Kind(val key: String) {
        WEIGHT("weight"),
        E1RM("e1rm"),
        REPS("reps"),
    }
}

/**
 * Personal records: whether a set just logged beat everything before it.
 *
 * The port of app/records.py, held to the same cases in fixtures/records.json:
 * heaviest weight, best estimated 1RM (Epley, as v_sets_enriched), and most
 * reps at this weight or heavier, which only counts where such a set exists.
 * A first set is no record. Warmups are neither checked nor counted.
 */
object Records {

    fun e1rm(weightKg: Double?, reps: Int?): Double? {
        if (weightKg == null || weightKg <= 0 || reps == null || reps !in 1..12) return null
        return round(weightKg * (1 + reps / 30.0) * 100) / 100
    }

    /** The records [new] sets over [before], in the order weight, e1rm, reps. */
    fun beaten(before: List<Pair<Double?, Int?>>, new: Pair<Double?, Int?>): List<Record> {
        val reps = new.second
        if (before.isEmpty() || reps == null) return emptyList()
        val weight = new.first ?: 0.0
        val found = mutableListOf<Record>()

        val heaviest = before.maxOf { it.first ?: 0.0 }
        if (weight > 0 && weight > heaviest) found += Record(Record.Kind.WEIGHT, weight, heaviest)

        val mine = e1rm(new.first, reps)
        val best = before.mapNotNull { e1rm(it.first, it.second) }.maxOrNull()
        if (mine != null && best != null && mine > best) found += Record(Record.Kind.E1RM, mine, best)

        val atOrAbove = before.filter { (it.first ?: 0.0) >= weight }.map { it.second ?: 0 }
        if (atOrAbove.isNotEmpty() && reps > atOrAbove.max()) {
            found += Record(Record.Kind.REPS, reps.toDouble(), atOrAbove.max().toDouble())
        }
        return found
    }

    /**
     * The exercise's name and the records this set broke, or null if there is
     * no such set. "Earlier" is an earlier day, or the same day and logged
     * first, as records.for_set.
     */
    fun forSet(db: Db, setId: Int): Pair<String, List<Record>>? {
        val id = setId.toString()
        data class Logged(val exerciseId: Int, val day: String, val name: String, val weight: Double?, val reps: Int?, val warmup: Boolean)
        val set =
            db.selectOne(
                """
                SELECT s.exercise_id, w.performed_on, e.name, s.weight_kg, s.reps, s.is_warmup
                FROM sets s
                JOIN workouts w ON w.id = s.workout_id
                JOIN exercises e ON e.id = s.exercise_id
                WHERE s.id = CAST(? AS INTEGER)
                """,
                listOf(id),
            ) {
                Logged(
                    it.int("exercise_id"),
                    it.string("performed_on"),
                    it.string("name"),
                    it.doubleOrNull("weight_kg"),
                    it.longOrNull("reps")?.toInt(),
                    it.boolean("is_warmup"),
                )
            } ?: return null
        if (set.warmup) return set.name to emptyList()

        val before =
            db.select(
                """
                SELECT s.weight_kg, s.reps
                FROM sets s JOIN workouts w ON w.id = s.workout_id
                WHERE s.exercise_id = CAST(? AS INTEGER) AND s.is_warmup = 0
                  AND s.id <> CAST(? AS INTEGER)
                  AND (w.performed_on < ? OR (w.performed_on = ? AND s.id < CAST(? AS INTEGER)))
                """,
                listOf(set.exerciseId.toString(), id, set.day, set.day, id),
            ) { it.doubleOrNull("weight_kg") to it.longOrNull("reps")?.toInt() }
        return set.name to beaten(before, set.weight to set.reps)
    }
}
